### PR TITLE
msg/async: fix outgoing_bl overflow and reset state_offset

### PR DESCRIPTION
**Pull request by Vicente-Cheng**
- we should reset state_offset when read done.

    - check outgoing_bl before we try to write a message.

    In some environments, network would temporily block and return EAGAIN.
    For async msgr, we would callback the write event directly, but that still
    increase the outgoing_bl.

    Think about this case, the sender is in congestion or network driver
    has some problems. The data appended to outgoing_bl and outgoing_bl
    is not consumed up-to-date immediately.
    That size of outgoing_bl will increase with time then overflow.
    The wrong outgoing_bl would cause some problems so we need to wait
    for outgoing_bl before we appended another message.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by fengchunsong** [comment]
If EAGAIN is returned, the socket eventfd is registered in AsyncConnection::_try_send:
  if (!open_write && is_queued()) {
    center->create_file_event(cs.fd(), EVENT_WRITABLE, write_handler);
    open_write = true;
  }

 When the fd is writable, the socket eventfd is called back and call  ProtocolV2::write_event again.
In this way, there is no problem with the existing mechanism.

**Comment by Vicente-Cheng** [comment]
hi @fengchunsong,
thanks for your reminders and tips.

yap, get EAGAIN (means `r = 0`) would directly trigger another callback to write_event.
In write_event we still append new data to outgoing_bl if the out_queue is not empty.
That still would make the outgoing_bl overflow in some situations.

I try to add a unit test to simulate this and verify overflow by adding the assert checking
Try to send the large message and block a while. I try to put around 10G data in 2 connections.
That would make outgoing_bl overflow.

**Comment by Vicente-Cheng** [comment]
hi @tchaikov ,

I am not sure who could(should) review it.
Could you help to review it?

**Comment by tchaikov** [comment]
will take a look next week.

**Comment by tchaikov** [comment]
@Vicente-Cheng pong. sorry for the latency. will do ASAP.

File: src/msg/async/ProtocolV1.cc [hunk 1]
```diff
@@ -315,6 +315,17 @@ void ProtocolV1::write_event() {
     auto start = ceph::mono_clock::now();
     bool more;
     do {
+
+      if (connection->outgoing_bl.length() > 0) {
+	r = connection->_try_send();
+	if (r < 0) {
+	  ldout(cct, 1) << __func__ << " pre-flush outgoing_bl fail." << dendl;
```
Comment (tchaikov) [inline src/msg/async/ProtocolV1.cc:322]: `AsyncConnection::_try_send()` already prints logging message when the underlying socket fails to send the buffer list, probably we can just drop this error message and put something like:
```c++
if (r = connection->_try_send(); r != 0) {
  // either fails to send or not all queued buffer is sent
  break;
}
```

File: src/msg/async/ProtocolV1.cc [hunk 2]
```diff
@@ -315,6 +315,17 @@ void ProtocolV1::write_event() {
     auto start = ceph::mono_clock::now();
     bool more;
     do {
+
+      if (connection->outgoing_bl.length() > 0) {
```
Comment (tchaikov) [inline src/msg/async/ProtocolV1.cc:319]:
```suggestion
      if (connection->is_queued()) {
```

File: src/common/options/global.yaml.in [hunk 3]
```diff
@@ -1262,6 +1262,12 @@ options:
   desc: Inject various internal delays to induce races (seconds)
   default: 0
   with_legacy: true
+- name: ms_inject_network_block
+  type: float
```
Comment (tchaikov) [inline src/common/options/global.yaml.in:1266]: how about:

1. rename it to `ms_inject_network_congestion`
1. use `uint`, to be consistent with `ms_inject_socket_failures`. as in this change `ms_inject_network_block` is actually a "ratio", while `ms_inject_socket_failures` is a period.

File: src/msg/async/AsyncConnection.cc [hunk 4]
```diff
@@ -362,6 +367,13 @@ void AsyncConnection::inject_delay() {
   }
 }
 
+bool AsyncConnection::inject_network_block() {
```
Comment (tchaikov) [inline src/msg/async/AsyncConnection.cc:370]:
```suggestion
bool AsyncConnection::inject_network_block() const {
```

File: src/msg/async/AsyncConnection.cc [hunk 5]
```diff
@@ -327,7 +328,11 @@ ssize_t AsyncConnection::_try_send(bool more)
   ceph_assert(center->in_thread());
   ldout(async_msgr->cct, 25) << __func__ << " cs.send " << outgoing_bl.length()
                              << " bytes" << dendl;
-  ssize_t r = cs.send(outgoing_bl, more);
+  // network block would make ::send return EAGAIN, that would make here looks
+  // like do not call cs.send() and r = 0
+  ssize_t r = 0;
+  if (!inject_network_block())
```
Comment (tchaikov) [inline src/msg/async/AsyncConnection.cc:334]:
```suggestion
  if (likely(!inject_network_block()) {
    r = cs.send(outgoing_bl, more);
  }
```

File: src/common/options/global.yaml.in (hunk 3 above)
Comment (Vicente-Cheng) [reply src/common/options/global.yaml.in:1266]: Hi, @tchaikov 
Sorry for the late reply, I am not sure of the above comment.

Do you mean that rename the config name and its type?
Then try to make it to be a ratio when  `ms_inject_socket_failures` is set up?

File: src/msg/async/ProtocolV1.cc (hunk 1 above)
Comment (Vicente-Cheng) [reply src/msg/async/ProtocolV1.cc]: sure, good suggestions.
already done with latest version

File: src/msg/async/ProtocolV1.cc (hunk 2 above)
Comment (Vicente-Cheng) [reply src/msg/async/ProtocolV1.cc]: thanks for reminder the related function
done with latest versions

File: src/common/options/global.yaml.in (hunk 3 above)
Comment (Vicente-Cheng) [reply src/common/options/global.yaml.in]: I rename this to `ms_inject_network_congestion`
That more like `ms_inject_socket_failures` and `ms_inject_network_congestion` means stuck with N times.

File: src/msg/async/AsyncConnection.cc (hunk 4 above)
Comment (Vicente-Cheng) [reply src/msg/async/AsyncConnection.cc]: done

File: src/msg/async/AsyncConnection.cc (hunk 5 above)
Comment (Vicente-Cheng) [reply src/msg/async/AsyncConnection.cc]: done

**Comment by Vicente-Cheng** [comment]
Hi @tchaikov, 
Already done for your suggestions.
Could you help to review it again?

Hi @rzarzynski,
Could you help to review it pr?

thanks!

File: src/msg/async/AsyncConnection.cc [hunk 6]
```diff
@@ -362,6 +368,13 @@ void AsyncConnection::inject_delay() {
   }
 }
 
+bool AsyncConnection::inject_network_congestion() const {
+  if (async_msgr->cct->_conf->ms_inject_network_congestion > 0 &&
+      rand() % async_msgr->cct->_conf->ms_inject_network_congestion != 0)
+    return true;
+  return false;
```
Comment (tchaikov) [inline src/msg/async/AsyncConnection.cc:375]:
```suggestion
  return (async_msgr->cct->_conf->ms_inject_network_congestion > 0 &&
          rand() % async_msgr->cct->_conf->ms_inject_network_congestion != 0);
```

File: src/test/msgr/test_msgr.cc [hunk 7]
```diff
@@ -1775,16 +1775,21 @@ class SyntheticWorkload {
   DummyAuthClientServer dummy_auth;
 
  public:
-  static const unsigned max_in_flight = 64;
-  static const unsigned max_connections = 128;
+  unsigned max_in_flight;
```
Comment (tchaikov) [inline src/test/msgr/test_msgr.cc:1778]: this member variable can, and should be marked `const`, i think.

File: src/test/msgr/test_msgr.cc [hunk 8]
```diff
@@ -1775,16 +1775,21 @@ class SyntheticWorkload {
...
+  unsigned max_connections;
```
Comment (tchaikov) [inline src/test/msgr/test_msgr.cc:1779]: ditto.

File: src/test/msgr/test_msgr.cc [hunk 9]
```diff
@@ -1775,16 +1775,21 @@ class SyntheticWorkload {
...
   static const unsigned max_message_len = 1024 * 1024 * 4;
 
   SyntheticWorkload(int servers, int clients, string type, int random_num,
-                    Messenger::Policy srv_policy, Messenger::Policy cli_policy)
+                    Messenger::Policy srv_policy, Messenger::Policy cli_policy,
+		    int _max_in_flight=64, int _max_connections=128)
```
Comment (tchaikov) [inline src/test/msgr/test_msgr.cc:1784]: might want to add spaces around `=`.

File: src/test/msgr/test_msgr.cc [hunk 10]
```diff
@@ -1918,6 +1923,31 @@ class SyntheticWorkload {
...
+    bufferlist bl;
+    string s("abcdefghijklmnopqrstuvwxyz");
+    for (int i = 0; i < 1024*256; i++)
+      bl.append(s);
+    // bl is around 6M
+    m->set_data(bl);
+    m->cmd = cmds;
+    m->set_priority(200);
+    // setup after connection is ready
+    if (inject_network_congestion && conn->is_connected())
```
Comment (tchaikov) [inline src/test/msgr/test_msgr.cc:1944]: please add curly brackets around the single-line statement for better readability.

File: src/test/msgr/test_msgr.cc [hunk 11]
```diff
@@ -1918,6 +1923,31 @@ class SyntheticWorkload {
...
+      g_ceph_context->_conf.set_val("ms_inject_network_congestion", "100");
+    else
+      g_ceph_context->_conf.set_val("ms_inject_network_congestion", "0");
```
Comment (tchaikov) [inline src/test/msgr/test_msgr.cc:1947]: ditto.

File: src/test/msgr/test_msgr.cc [hunk 12]
```diff
@@ -2196,6 +2226,35 @@ TEST_P(MessengerTest, SyntheticInjectTest4) {
...
 
+// This is test for network block, means ::send return EAGAIN
+TEST_P(MessengerTest, SyntheticInjectTest5) {
+  SyntheticWorkload test_msg(1, 8, GetParam(), 100,
+                             Messenger::Policy::stateful_server(0),
+                             Messenger::Policy::lossless_client(0),
+			     64, 2);
+  bool simulate_network_congestion = true;
+  for (int i = 0; i < 2; ++i) {
+    if (!(i % 10)) lderr(g_ceph_context) << "seeding connection " << i << dendl;
```
Comment (tchaikov) [inline src/test/msgr/test_msgr.cc:2237]: what is this for?

File: src/test/msgr/test_msgr.cc [hunk 13]
```diff
@@ -2196,6 +2226,35 @@ TEST_P(MessengerTest, SyntheticInjectTest4) {
...
+    test_msg.generate_connection();
+  }
+  for (int i = 0; i < 5000; ++i) {
+    if (!(i % 10)) {
+      lderr(g_ceph_context) << "Op " << i << ": " << dendl;
```
Comment (tchaikov) [inline src/test/msgr/test_msgr.cc:2242]: IMHO, `lderr()` is only for logging errors.

File: src/test/msgr/test_msgr.cc [hunk 14]
```diff
@@ -2196,6 +2226,35 @@ TEST_P(MessengerTest, SyntheticInjectTest4) {
...
+    }
+    if (i < 1600) {
+      // means that we would stuck 1600 * 6M (9.6G) around with 2 connections
+      test_msg.send_large_message(simulate_network_congestion);
+    } else {
+      simulate_network_congestion = false;
+      test_msg.send_large_message(simulate_network_congestion);
+    }
+
+
```
Comment (tchaikov) [inline src/test/msgr/test_msgr.cc:2253]: please remove unnecessary vertical spaces.

File: src/test/msgr/test_msgr.cc (hunk 12 above)
Comment (Vicente-Cheng) [reply src/test/msgr/test_msgr.cc]: redundant code, already remove it.

File: src/test/msgr/test_msgr.cc (hunk 13 above)
Comment (Vicente-Cheng) [reply src/test/msgr/test_msgr.cc:2240]: do nothing because I remove this line :p

File: src/test/msgr/test_msgr.cc (hunk 14 above)
Comment (Vicente-Cheng) [reply src/test/msgr/test_msgr.cc]: done

File: src/msg/async/AsyncConnection.cc (hunk 6 above)
Comment (Vicente-Cheng) [reply src/msg/async/AsyncConnection.cc]: done, thanks for clarifying

File: src/test/msgr/test_msgr.cc (hunk 7 above)
Comment (Vicente-Cheng) [reply src/test/msgr/test_msgr.cc]: yes, already make it const.

File: src/test/msgr/test_msgr.cc (hunk 9 above)
Comment (Vicente-Cheng) [reply src/test/msgr/test_msgr.cc]: done

File: src/test/msgr/test_msgr.cc (hunk 10 above)
Comment (Vicente-Cheng) [reply src/test/msgr/test_msgr.cc]: done

**Comment by Vicente-Cheng** [comment]
Hi @tchaikov, 
already done for the above suggestions.

please review it again.
thanks!

File: src/test/msgr/test_msgr.cc (hunk 13 above)
Comment (Vicente-Cheng) [reply src/test/msgr/test_msgr.cc]: hmm, just make the same with the above logging method.

here logging the statement, using normal `ldout` would be fine?

File: src/msg/async/ProtocolV2.cc [hunk 15]
```diff
@@ -653,6 +653,14 @@ void ProtocolV2::write_event() {
     auto start = ceph::mono_clock::now();
     bool more;
     do {
+
```
Comment (tchaikov) [inline src/msg/async/ProtocolV2.cc:656]: could you remove the empty line above?

**Review by tchaikov — APPROVED** [review]
modulo a nit, LGTM in general.

File: src/msg/async/ProtocolV2.cc (hunk 15 above)
Comment (Vicente-Cheng) [reply src/msg/async/ProtocolV2.cc]: sure, already remove the empty line.

**Comment by Vicente-Cheng** [comment]
> modulo a nit, LGTM in general.

hi @tchaikov ,

fix nits, thanks for the review! :)

**Comment by yuriw** [comment]
The tests run analysis by @kamoltat 

`Run analysis: wip-yuri-testing-2022-02-09-1607
Jobid: [6677097, 6677098, 6677102, 6677106]
Description: rados/rook/smoke/{0-distro/ubuntu_20.04 0-kubeadm 0-nvme-loop 1-rook 2-workload ...
Error: rook/smoke: Failed to write to /dev/nvme-fabrics: Input/output error
Tracker: https://tracker.ceph.com/issues/54094#change-210010

Jobid: [6677099, 6677103] - related to a PR in this batch https://github.com/ceph/ceph/pull/44911#issuecomment-1036617782
Description: rados/cephadm/workunits/{agent/on mon_election/connectivity task/test_nfs}
Error: 
2022-02-11T19:05:26.757 DEBUG:teuthology.orchestra.run.smithi090:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph nfs cluster create test
2022-02-11T19:05:27.080 INFO:teuthology.orchestra.run.smithi090.stderr:Error EPERM: osd pool application enable failed: unrecognized pool '.nfs' retval: -2
2022-02-11T19:05:27.083 DEBUG:teuthology.orchestra.run:got remote process result: 1
2022-02-11T19:05:27.602 INFO:tasks.cephfs_test_runner:test_cluster_info (tasks.cephfs.test_nfs.TestNFS) ... ERROR
Tracker: https://tracker.ceph.com/issues/54262?next_issue_id=54257`

**Pull request merged**